lease/lessor_bench_test.go

```diff
@@ -15,112 +15,207 @@
-func BenchmarkLessorRevoke100(b *testing.B)     { benchmarkLessorRevoke(100, b) }
-func BenchmarkLessorRevoke1000(b *testing.B)    { benchmarkLessorRevoke(1000, b) }
-func BenchmarkLessorRevoke10000(b *testing.B)   { benchmarkLessorRevoke(10000, b) }
-func BenchmarkLessorRevoke100000(b *testing.B)  { benchmarkLessorRevoke(100000, b) }
-func BenchmarkLessorRevoke1000000(b *testing.B) { benchmarkLessorRevoke(1000000, b) }
-
-func benchmarkLessorFindExpired(size int, b *testing.B) {
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	le.Promote(0)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+func BenchmarkLessorGrant1000(b *testing.B)   { benchmarkLessorGrant(1000, b) }
+func BenchmarkLessorGrant100000(b *testing.B) { benchmarkLessorGrant(100000, b) }
+
+func BenchmarkLessorRevoke1000(b *testing.B)   { benchmarkLessorRevoke(1000, b) }
+func BenchmarkLessorRevoke100000(b *testing.B) { benchmarkLessorRevoke(100000, b) }
+
+func BenchmarkLessorRenew1000(b *testing.B)   { benchmarkLessorRenew(1000, b) }
+func BenchmarkLessorRenew100000(b *testing.B) { benchmarkLessorRenew(100000, b) }
+
+// Use findExpired10000 replace findExpired1000, which takes too long.
+func BenchmarkLessorFindExpired10000(b *testing.B)  { benchmarkLessorFindExpired(10000, b) }
+func BenchmarkLessorFindExpired100000(b *testing.B) { benchmarkLessorFindExpired(100000, b) }
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
+
+const (
+	// minTTL keep lease will not auto expire in benchmark
+	minTTL = 1000
+	// maxTTL control repeat probability of ttls
+	maxTTL = 2000
+)
+
+func randomTTL(n int, min, max int64) (out []int64) {
+	for i := 0; i < n; i++ {
+		out = append(out, rand.Int63n(max-min)+min)
 	}
-	le.mu.Lock() //Stop the findExpiredLeases call in the runloop
+	return out
+}
+
+// demote lessor from being the primary, but don't change any lease's expiry
+func demote(le *lessor) {
+	le.mu.Lock()
 	defer le.mu.Unlock()
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		le.findExpiredLeases(1000)
-	}
+	close(le.demotec)
+	le.demotec = nil
 }
 
-func benchmarkLessorGrant(size int, b *testing.B) {
+// return new lessor and tearDown to release resource
+func setUp() (le *lessor, tearDown func()) {
 	lg := zap.NewNop()
 	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
-	}
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		le.Grant(LeaseID(i+size), int64(100+i+size))
+	// MinLeaseTTL is negative, so we can grant expired lease in benchmark.
+	// ExpiredLeasesRetryInterval should small, so benchmark of findExpired will recheck expired lease.
+	le = newLessor(lg, be, LessorConfig{MinLeaseTTL: -100, ExpiredLeasesRetryInterval: 10 * time.Microsecond})
```

Comment:
-100 is not enough for the TTLs in benchmarkLessorFindExpired.

Reply:
Thanks